sandbox/cgroup/kill.go

```diff
@@ -0,0 +1,128 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"path/filepath"
+	"syscall"
+)
+
+// KillSnapProcesses sends a signal to all the processes belonging to
+// a given snap.
+//
+// A correct implementation is picked depending on cgroup v1 or v2 use in the
+// system. When cgroup v1 is detected, the call will directly act on the freezer
+// group created when a snap process was started, while with v2 the call will
+// act on all tracking groups of a snap.
+//
+// XXX: It is important to note that the implementation for v2 is racy because
+// processes can be killed externally even when the cgroup is frozen and have
+// their pid reused. Also, v2 freezing support is a no-op on kernels before
+// 5.2 which means new processes can keep getting spawned while killing
+// pids read earlier.
+var KillSnapProcesses = killSnapProcessesImplV1
+
+func pickKillV1Impl() {
```

Comment:
I think it might be cleaner to just have one toggle v1, v2, fake and have it flip all the functions related to cgroups, so that it's not possible to skew.